src/cli/Hibp.cpp

```diff
@@ -0,0 +1,221 @@
+/*
+ *  Copyright (C) 2019 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <cstdlib>
+#include <cstring>
+#include <stdio.h>
+
+#include "Hibp.h"
+#include "cli/Utils.h"
+
+#include <QCommandLineParser>
+#include <QFile>
+
+#include "cli/TextStream.h"
+#include "core/Database.h"
+#include "core/Entry.h"
+#include "core/Group.h"
+#include "core/Tools.h"
+#include "crypto/CryptoHash.h"
+
+namespace
+{
+
+    struct Sha1Hash
+    {
+        uchar m_data[20];
+
+        Sha1Hash()
+        {
+        }
+
+        Sha1Hash(const QByteArray& a)
+        {
+            Q_ASSERT(a.size() == sizeof(m_data));
+            std::memcpy(&m_data[0], a.data(), sizeof(m_data));
+        }
+
+        bool operator==(const Sha1Hash& other) const
+        {
+            return std::memcmp(&m_data[0], &other.m_data[0], sizeof(m_data)) == 0;
+        }
+    };
+
+    uint qHash(const Sha1Hash& b)
+    {
+        uint result;
+        std::memcpy(&result, &b.m_data[0], sizeof(result));
+        return result;
+    }
+
+    bool parseHibpLine(QFile& file, Sha1Hash& sha1, size_t& numPwns, bool& atEnd, QString& error)
+    {
+        auto failedTo = [&](const QString& action) {
+            error = QObject::tr("failed to %1").arg(action);
+            return false;
+        };
+
+        auto failedToIo = [&](const QString& action) {
+            error = QObject::tr("failed to %1: %2").arg(action).arg(file.errorString());
+            return false;
+        };
+
+        atEnd = false;
+
+        char hexSha1[sizeof(Sha1Hash::m_data) * 2];
```

Comment:
Use QByteArray and its size() method (which is O(1) instead of O(n) like sizeof is).

Reply:
`sizeof` is not O(n). It gets evaluated at compile time.

Reply:
Only true for non-dynamic arrays, which at the time of evaluation you cannot guarantee, since the declaration may change (however unlikely that may be). Generally, do not use plain C arrays.